// elsass/util/string.js
/**
 * 字符串相关处理模块
 */

let StringUtils = {
  /**
   * 去除所有空格
   * @param str
   */
  trimAll (str) {
    console.log(str)
    return str.replace(/\s+/g, '')
  },
  /**
   * 去除两头的空格
   * @param str
   */
  trim (str) {
    return str.replace(/^\s+|\s+$/g, '')
  },
  /**
   * 去除左边的空格
   * @param str
   * @return {*}
   */
  ltrim (str) {
    return str.replace(/^\s*/, '')
  },
  /**
   * 去除右边的空格
   * @param str
   * @return {*}
   */
  rtrim (str) {
    return str.replace(/(\s*$)/g, '')
  },
  /**
   * 判断字符串是否为空
   * @param str
   * @return {boolean}
   */
  isBlank (str) {
    if (str === undefined || str === null || this.trimAll(str) === '') {
      return true
    }
    return false
  },
  /**
   * 判断字符串是否超过len个字节
   * @param str
   * @param len
   * @return {boolean}
   */
  overLen (str, len) {
    if (!len || !(typeof(len) === 'number')) {
      return false
    }
    // 中文字占两个字节，英文占一个字节。
    // 把中文字替换成两个字母。然后计算长度
    let _str = str.replace(/[\u4e00-\u9fa5]/g, "aa")
    return _str.length > len
  }
}

export default StringUtils
